src/integrations/devicecontext.ts

```diff
@@ -19,29 +20,119 @@ export class DeviceContext implements Integration {
       if (!self) {
         return event;
       }
 
+      let nativeContexts: NativeDeviceContextsResponse | null = null;
       try {
-        const contexts = await NATIVE.fetchNativeDeviceContexts();
-        const context = (contexts['context'] as Contexts);
-
-        event.contexts = { ...context, ...event.contexts };
-        if ('user' in contexts) {
-          const user = contexts['user'];
-          if (!event.user) {
-            event.user = { ...user };
-          }
-        }
+        nativeContexts = await NATIVE.fetchNativeDeviceContexts();
       } catch (e) {
         logger.log(`Failed to get device context from native: ${e}`);
       }
+      if (!nativeContexts) {
+        return event;
+      }
+
+      if (nativeContexts.contexts) {
+        event.contexts = { ...nativeContexts.contexts, ...event.contexts };
+        if (nativeContexts.contexts.app) {
+          event.contexts.app = { ...nativeContexts.contexts.app, ...event.contexts.app };
+        }
+      }
+      if (nativeContexts.user) {
+        const user = nativeContexts.user;
+        if (!event.user) {
+          event.user = { ...user };
+        }
+      }
+      const nativeTags = nativeContexts.tags;
+      if (nativeTags) {
+        event.tags = { ...nativeTags, ...event.tags };
+      }
+
+      const nativeExtra = nativeContexts.extra;
+      if (nativeExtra) {
+        event.extra = { ...nativeExtra, ...event.extra };
+      }
+
+      const nativeFingerprint = nativeContexts.fingerprint;
+      if (nativeFingerprint) {
+        event.fingerprint = (event.fingerprint ?? []).concat(
+          nativeFingerprint.filter(item => (event.fingerprint ?? []).indexOf(item) < 0),
+        );
+      }
+
+      if (!event.level && nativeContexts.level) {
+        event.level = severityLevelFromString(nativeContexts.level);
+      }
+
+      const nativeEnvironment = nativeContexts.environment;
+      if (!event.environment && nativeEnvironment) {
+        event.environment = nativeEnvironment;
+      }
+
+      const nativeBreadcrumbs = Array.isArray(nativeContexts.breadcrumbs)
+      ? nativeContexts.breadcrumbs.map(breadcrumbFromObject)
+      : undefined;
+      if (nativeBreadcrumbs) {
+        if (event.breadcrumbs && event.breadcrumbs.length !== nativeBreadcrumbs.length) {
+          event.breadcrumbs = this._mergeUniqueBreadcrumbs(event.breadcrumbs, nativeBreadcrumbs);
+        }
+        else {
+          event.breadcrumbs = nativeBreadcrumbs;
+        }
+      }
 
       return event;
     });
   }
+
+  /**
+ * Merges two groups of ordered breadcrumbs and removes any duplication that may
+ * happen between them.
+ * @param jsList The first group of breadcrumbs from the JavaScript layer.
+ * @param nativeList The second group of breadcrumbs from the native layer.
+ * @returns An array of unique breadcrumbs merged from both lists.
+   */
+  private _mergeUniqueBreadcrumbs(jsList: Breadcrumb[], nativeList: Breadcrumb[]): Breadcrumb[] {
```

Comment:
Thank you for adding test for this method, it makes it super easy to understand how it would behave.

m: I'm not sure if we can be sure that the breadcrumbs are in the same order in both arrays.

If I'm looking at the correctly in case they would not be in the same order the duplicates would not be removed. Creating possible double the size.

We should also respect the `maxBreadcrumbs` size.

Reply:
I have added tests for the maxBreadcrumb and also for out of order breadcrumbs